### ISSUE TITLE
Boot Loop After Changing Net VM

### ISSUE THREAD

**Issue by straussmaus** [QubesOS/qubes-issues#6083]
**Qubes OS version**

4.0.1

**Affected component(s) or functionality**

Booting

**Brief summary**
I changed my NetVM to a clone of one that should have the MAC changer functionality I got from the Qubes docs.

I started this NetVM and the system rebooted itself.

I entered my encryption password, and the system reboots itself.

I have a feeling it is because I created this MAC changer NetVM on a different system, where the network interfaces had a different device ID.  And now on this new computer, giving it control of Device X is actually not giving it the network controllers but something else.  If so, then the bug is that there was no warning about something that could cause a boot loop.  So once we find the root source, I can help propose a solution.

I can get console messages but I have one system so I have to write and type them.  So I'm hoping you can tell me what sort of messages to look for, so I can add them to this Issue.

Also, is there a "Safe Boot" type option somewhere, and if so how can I use that to change which is the default NetVM?  Then I could boot and fix the problem.

**To Reproduce**

**Expected behavior**

**Actual behavior**

**Additional context**

See above please.

**Solutions you've tried**

I'm trying to get into a safe mode, so I've tried all the magic keys I can think of.

I am also looking for what files on the system control which is the default NetVM.  Maybe I can boot the system to Knoppix, decrypt the drive, and manually fix it?

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

I haven't found anything related.

**Comment by brendanhoar** [comment]
Does your bios allow you to disable vt-d? That might prevent sys-net from starting and you could then disable auto-start for that VM, shut down, reboot into bios, enable vt-d then reboot and create a new sys-net.

**Comment by straussmaus** [comment]
I don't know what vt-d is but I think the answer is no.  I use coreboot and the capabilities are very limited.  What is vt-d?  Is it known by other names?  I could look but I really doubt it.  But thank you for the suggestion.

**Comment by straussmaus** [comment]
@andrewdavidwong 

I agree this needs diagnosis.  But also, a workaround.

Is there a flag to add to this so we can find a workaround?  ie  - a command I can run from another linux machine, after mounting the encrypted disk, to change the NetVM?

Is there a doc that explains this that I have not seen maybe?

Thank you.

**Comment by l1am9111** [comment]
It is possible to stop Qubes autostarting any VMs by adding `qubes.skip_autostart` to kernel command line.

**Comment by andrewdavidwong** [comment]
> It is possible to stop Qubes autostarting any VMs by adding `qubes.skip_autostart` to kernel command line.

This sounds like a potential workaround worth trying. I am not aware of anything else.

**Comment by straussmaus** [comment]
I solved the problem because I had the old system available.  So I put the drive in the old system, booted, fixed things, put it in the new machine.

But this is a potentially big problem for anyone who backs up and restores to a new or different computer.  Or who has an archival copy of their data, loses their computer, and gets a new one.  So a good fix is important.

@l1am9111 Where do I enter this kernel parameter.  I have not been able to get a "boot" prompt to add kernal parameters.  (Never needed it before on Qubes.)

**Comment by andrewdavidwong** [comment]
> But this is a potentially big problem for anyone who backs up and restores to a new or different computer.

What method was used to back up the VMs on the old machine and restore them on the new machine?

**Comment by straussmaus** [comment]
> What method was used to back up the VMs on the old machine and restore them on the new machine?

I used the Qube Manager System > Backup qubes and then restored the same way.

**Comment by straussmaus** [comment]
@l1am9111 

> It is possible to stop Qubes autostarting any VMs by adding qubes.skip_autostart to kernel command line.

Where do I enter this command?  Where is the Qubes kernel command line?

**Comment by unman** [comment]
On Wed, Dec 09, 2020 at 03:17:49AM -0800, straussmaus wrote:
> @l1am9111 
> 
> > It is possible to stop Qubes autostarting any VMs by adding qubes.skip_autostart to kernel command line.
> 
> Where do I enter this command?  Where is the Qubes kernel command line?
> 
qubes-issues is really not the place for support questions. You
would get faster responses and your questions would be seen by more
people in the mailing lists or in the forum. It would also have the
benefit that other users would be more likely to find your question and
the answer.

If you use grub, then you can interrupt the boot process, and edit the
kernel command line - the editor will allow you to add that to the
kernel line, and then you can boot. This is standard stuff, and well
documented.
If you use UEFI? Ask for support in the right place.

And before you say it, I would tell you this in person.

**Comment by straussmaus** [comment]
> And before you say it, I would tell you this in person.

Well there is nothing wrong with what you said this time. You are incorrect, but you are not rude this time. :)

Thank you for clarifying that @l1am9111  is referring to a standard kernel command line, though he did not explain how to reach it in Qubes.  Qubes seems to install with defaults that do not allow for accessing GRUB command line?  But at least you cleared up that this is what he meant, and that is helpful.  Thank you.

**Comment by straussmaus** [comment]
I noticed that this has a hardware support tag.  I wonder if that is appropriate.  The issue may have a software fix.  When restoring, maybe it makes sense for there to be a checkbox to restore hardware devices or not with the warning:

`WARNING: if your target hardware is different from source hardware, this may cause errors!"`

**Comment by DemiMarie** [comment]
Agree.  I think that *not* restoring assigned PCI devices should be the default, as this is a security concern.

**Comment by marmarek** [comment]
Re security concern: the qube itself have no control over its devices, that comes from backup metadata (qubes.xml). And this (as all the other parts of the backup) is obviously integrity protected. In case of migrating to another hardware this indeed may match into different physical devices, but in practice you'll need to hit all of them for given qube to have it started. And there are not that many devices that a) can be assigned at all, _and_ b) would allow to do something nasty _and_ c) collide with something innocent on some other machine _and_ d) other devices assigned to a qube are also present.

As for restoring devices in general: the main purpose of the backup is to restore the system to the state saved there as closely as possible. This needs to let you restore your system to the state where you can use it again, without spending half a day trying to figure out what needs to be set manually to get network working... There are still some parts of the system that are not preserved in the backup and this is actually something to be fixed.
That said, there are indeed cases where restoring devices 1:1 is not desired. And yes, having an option for that is a good idea.
One thing we have considered in the past is an option to dynamically assign devices based on a type (https://github.com/QubesOS/qubes-issues/issues/214) - thing like "assign all network devices" instead of "assign device 00:19.0". That might solve similar issue, but I don't think it will cover all the cases.

**Comment by straussmaus** [comment]
I think @DemiMarie means the security concern is that a user could end up with devices mapped unexpectedly, and therefore have a security breach by giving the wrong Qube access to a network or USB device.  Is that correct DM?

Anyway I don't know how many is "not that many devices" but in my case it was just a matter of migrating from one laptop to the next model of that laptop.  These were both similar and also different enough to have the problem, and upgrading to a similar but newer system can't be *so* unusual?

I agree that "assign all network devices" would not cover all cases.  

I think a generic checkbox to not import settings, or maybe just a separate one for sys-net and sys-usb because they are the ones where the error is most likely to cause major problems, could be a solution, if part a *restore* function.  No reason not to back up settings, but during restore maybe "Do not import Devices to restored Qube (especially useful when migrating to new target hardware)"?

**Comment by DemiMarie** [comment]
> I think @DemiMarie means the security concern is that a user could end up with devices mapped unexpectedly, and therefore have a security breach by giving the wrong Qube access to a network or USB device. Is that correct DM?

Or worse.  There are some devices that are basically Game Over.  Things like integrated graphics, internal storage, and probably some obscure devices I haven’t thought of, although @marmarek would know.

That said, @marmarek is also correct in that an attacker has no way to make this happen.  The best they can do is opportunistically exploit the situation if it does arise.  One is far more likely to wind up with an unbootable machine, as you did.  Nevertheless, that is still a rather severe reliability problem.

We also need to be able to restore without device assignments.  Even if one is restoring the backup onto the same system one made it on, it is still possible to perform the restoration outside of dom0 via the Admin API.  And we need to handle the situation where the qube performing the registration is not allowed to assign devices.  It is also quite possible that one of the devices on the old system might not even exist on the new system, and that should not cause a restore to fail.

> I think a generic checkbox to not import settings, or maybe just a separate one for sys-net and sys-usb because they are the ones where the error is most likely to cause major problems, could be a solution, if part a _restore_ function. No reason not to back up settings, but during restore maybe "Do not import Devices to restored Qube (especially useful when migrating to new target hardware)"?

I absolutely agree.  Including the device assignment information in the backup is harmless, but automatically restoring them is risky.  Since restoring from a backup is likely to be relatively rare, and since the number of devices on most machines is farely small, we could prompt the user for each device.  To make this possible, I would like for the backup to include the human-readable names, as well as the machine-readable device IDs.

@marmarek: Do PCI devices have some sort of unique identifier?  If a given device on the target is the same as that of the source, then restoring the assignment makes sense (though I would like to be able to opt-out).  Otherwise, not restoring the assignment seems like the safer option.  Alternatively, we could restore the assignments in some sort of disabled state, where they are available in the GUI and CLI but do nothing until activated.

**Comment by marmarek** [comment]
> We also need to be able to restore without device assignments. Even if one is restoring the backup onto the same system one made it on, it is still possible to perform the restoration outside of dom0 via the Admin API. And we need to handle the situation where the qube performing the registration is not allowed to assign devices. It is also quite possible that one of the devices on the old system might not even exist on the new system, and that should not cause a restore to fail.

Yes, this is already handled - generally errors in restoring individual VM property or other metadata do not interrupt the restore process and are simply reported. Here is also the policy for restoring a backup using a VM (which can be DisposableVM):  https://github.com/QubesOS/qubes-core-admin/blob/master/qubes-rpc-policy/85-admin-backup-restore.policy

> Do PCI devices have some sort of unique identifier?

Not really. The way the system identifies a device is BDF (bus-device-function), which basically describes physical/logical slot where device is plugged in. If you unplug a device and plug another one in the same slot, it will get the same BDF (well, the number of functions may differ, but lets ignore it for now). Other than that, there is also Vendor ID and Device ID - those are not unique at all - if you have multiple instances of some device, all will have the same VID:DID pair. But they do identify device model uniquely.
So, what we can do, is to record those identifiers, and when attaching a device verify if that is still the same VID:DID - if not, do not allow to attach the device (report as failed attach).

**Comment by DemiMarie** [comment]
> Not really. The way the system identifies a device is BDF (bus-device-function), which basically describes physical/logical slot where device is plugged in. If you unplug a device and plug another one in the same slot, it will get the same BDF (well, the number of functions may differ, but lets ignore it for now). Other than that, there is also Vendor ID and Device ID - those are not unique at all - if you have multiple instances of some device, all will have the same VID:DID pair. But they do identify device model uniquely.
> So, what we can do, is to record those identifiers, and when attaching a device verify if that is still the same VID:DID - if not, do not allow to attach the device (report as failed attach).

That seems like the best option by far.  Ideally, we could have some sort of mapping from VID:DID to device types, so as to offer suggestions.

**Comment by straussmaus** [comment]
> So, what we can do, is to record those identifiers, and when attaching a device verify if that is still the same VID:DID - if not, do not allow to attach the device (report as failed attach).

That seems dangerous from a privacy perspective.  There could be a situation where a person changes to new hardware specifically because it has been identified.  Anyway this would mean that if a backup was compromised, the adversary would have the data, and be able to tie that data to a hardware device (and possibly purchase records, etc.)

**Comment by DemiMarie** [comment]
> > So, what we can do, is to record those identifiers, and when attaching a device verify if that is still the same VID:DID - if not, do not allow to attach the device (report as failed attach).
> 
> That seems dangerous from a privacy perspective. There could be a situation where a person changes to new hardware specifically because it has been identified. Anyway this would mean that if a backup was compromised, the adversary would have the data, and be able to tie that data to a hardware device (and possibly purchase records, etc.)

Could this be optional?  Note that this data is only sufficient to identify a given make and model, not an individual piece of hardware.  Also, it would be encrypted along with the rest of the backups, and if the encryption was broken, there may well be other ways to determine what kind of hardware the backup was made by.  In any case, I think that blindly assigning a PCI device is just too dangerous.

**Comment by brendanhoar** [comment]
> Could this be optional?  Note that this data is only sufficient to identify a given make and model, not an individual piece of hardware.  Also, it would be encrypted along with the rest of the backups, and if the encryption was broken, there may well be other ways to determine what kind of hardware the backup was made by.  In any case, I think that blindly assigning a PCI device is just too dangerous.

My thought reading this yesterday was that each VM has an "enable any assigned PCIe devices" flag/switch. 

Normally HVMs when assigned PCIe device through the standard tools or interface have this flag enabled. If the device is attached through a restore, however, the flag is unset.

When starting a VM with devices assigned but the flag unset, a reminder notification is given that the device is associated with the VM but should be validated by the user and then have the flag be set again so that it can be used by the VM.

An alternative method would be a "block associated PCIe devices" or "please affirmatively validate you wanted these assigned" flag. E.g. all of the above, just reverse set and unset in all scenarios.

B

**Comment by marmarek** [comment]
> That seems dangerous from a privacy perspective. There could be a situation where a person changes to new hardware specifically because it has been identified. Anyway this would mean that if a backup was compromised, the adversary would have the data, and be able to tie that data to a hardware device (and possibly purchase records, etc.)

I think it is very unrealistic expectation that a backup of a VM that had direct hardware access (sys-net, sys-usb) would not allow to identify (at least make and model of) that hardware. At the very least, you have a bunch of logs, network manager configuration (likely including real MAC address) and possibly dozen more. As @DemiMarie said, the backup is encrypted and this encryption includes VMs metadata. If you worry about adversary that has access to a plaintext of said backup, then... you can simply not include sys-net and sys-usb in the backup (I still think if an adversary has access to a plaintext of your full backup, you are basically screwed, but that's offtopic here).

**Comment by straussmaus** [comment]
@marmarek 

> I think it is very unrealistic expectation that a backup of a VM...

Okay, point taken.

But there is still a security concern.  Imagine your backup is not compromised but you are careful to only use certain USB devices for certain Qubes.  Then you restore to new hardware and because the USB assignments do not match your new hardware, you could easily compromise a Qube or a USB drive, etc.

I'm thinking of a USB drive that you use carefully on your system because you also bring it to the local print shop and plug it into the public Windows computer to print things with.

Obviously it's good practice not to have "dirty" drives plugged in on your first run.  But what about a situation where you have a Qube that you don't use often, so you've forgotten about this issue by the time you start it again.  (In that case you're more likely to infect a clean or "air-gapped" drive.)

I like @DemiMarie's suggestion:

> Agree. I think that not restoring assigned PCI devices should be the default, as this is a security concern.
